components/brave_rewards/core/common/signer.cc

```diff
@@ -0,0 +1,85 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_rewards/core/common/signer.h"
+
+#include <openssl/evp.h>
+#include <openssl/hkdf.h>
+#include <utility>
+
+#include "crypto/random.h"
+#include "tweetnacl.h"  // NOLINT
+
+namespace {
+
+const uint8_t kHkdfSalt[] = {
+    126, 244, 99,  158, 51,  68,  253, 80,  133, 183, 51,  180, 77,
+    62,  74,  252, 62,  106, 96,  125, 241, 110, 134, 87,  190, 208,
+    158, 84,  125, 69,  246, 207, 162, 247, 107, 172, 37,  34,  53,
+    246, 105, 20,  215, 5,   248, 154, 179, 191, 46,  17,  6,   72,
+    210, 91,  10,  169, 145, 248, 22,  147, 117, 24,  105, 12};
+
+const int kSeedLength = 32;
+
+const int kSaltLength = 64;
+
+}  // namespace
+
+namespace brave_rewards::internal {
+
+Signer::Signer(std::vector<uint8_t> public_key, std::vector<uint8_t> secret_key)
+    : public_key_(std::move(public_key)), secret_key_(std::move(secret_key)) {
+  DCHECK_EQ(public_key_.size(),
+            static_cast<size_t>(crypto_sign_PUBLICKEYBYTES));
+  DCHECK_EQ(secret_key_.size(),
+            static_cast<size_t>(crypto_sign_SECRETKEYBYTES));
+}
+
+Signer::~Signer() = default;
+
+Signer::Signer(const Signer&) = default;
+Signer& Signer::operator=(const Signer&) = default;
+
+std::optional<Signer> Signer::FromRecoverySeed(
+    const std::vector<uint8_t>& recovery_seed) {
+  if (recovery_seed.size() != kSeedLength) {
+    return std::nullopt;
+  }
+
+  std::vector<uint8_t> secret_key(kSeedLength);
+  const uint8_t info[] = {0};
+  int hkdf_res = HKDF(secret_key.data(), kSeedLength, EVP_sha512(),
+                      recovery_seed.data(), kSeedLength, kHkdfSalt, kSaltLength,
+                      info, sizeof(info) / sizeof(info[0]));
+
+  DCHECK(hkdf_res);
+
+  std::vector<uint8_t> public_key(crypto_sign_PUBLICKEYBYTES);
+  secret_key.resize(crypto_sign_SECRETKEYBYTES);
+
+  crypto_sign_keypair(public_key.data(), secret_key.data(), /* seeded */ 1);
+
+  return Signer(std::move(public_key), std::move(secret_key));
+}
+
+std::vector<uint8_t> Signer::GenerateRecoverySeed() {
```

Comment:
Do we plan to use this for anything but the single key? This isn't really a seed phrase in the since that it's usable and storable by a person since it's just RandBytes. I'm wondering if we need this value at all or if it makes sense to just generate the private key and store that rather than needing to keep two secrets.

Reply:
I'm not sure. It might make sense to revisit this at a later time; moving away from the "recovery seed" approach would be a significant engineering effort (it is also used by the `brave_ads` code).

Reply:
Ahh ok, that answers my question then that this is operating as a piece of entropy for multiple places. In that case, not worth changing this.